src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -0,0 +1,307 @@
+            var roslynTrigger = Helpers.GetRoslynTrigger(trigger, triggerLocation);
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                roslynTrigger).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, AsyncCompletionData.CompletionFilter>();
+
+            var itemsBuilder = new ArrayBuilder<VSCompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+
+                // Have to store the snapshot to reuse it in some projections related scenarios
+                // where data and session in further calls are able to provide other snapshots.
+                item.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                itemsBuilder.Add(item);
+            }
+
+            var items = itemsBuilder.ToImmutableAndFree();
+
+            var suggestionItemOptions = completionList.SuggestionModeItem != null
+                    ? new AsyncCompletionData.SuggestionItemOptions(
+                        completionList.SuggestionModeItem.DisplayText,
+                        completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                            ? description
+                            : string.Empty)
+                    : null;
+
+            // This is a code supporting legacy completion scenarios:
+            // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
+            session.Properties.AddProperty(HasSuggestionItemOptions, suggestionItemOptions != null);
+
+            session.Properties.AddProperty(InitialTriggerKind, roslynTrigger.Kind);
+            var excludedCommitCharacters = GetExcludedCommitCharacters(completionList.Items);
+            if (excludedCommitCharacters.Length > 0)
+            {
+                session.Properties.AddProperty(ExcludedCommitCharacters, excludedCommitCharacters);
+            }
+
+            return new AsyncCompletionData.CompletionContext(
+                items,
+                suggestionItemOptions,
+                suggestionItemOptions == null 
+                ? AsyncCompletionData.InitialSelectionHint.RegularSelection 
+                : AsyncCompletionData.InitialSelectionHint.SoftSelection);
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, VSCompletionItem item, CancellationToken cancellationToken)
+        {
+            // The method is async but is called in the UI thread.
+            AssertIsForeground();
```

Comment:
1. i'm not sure i get this assert.   we're basically saying, if the caller changes the thread, we'll throw.  Is that really what we want here? I don't see anything here that needs to eb on the UI thread.
2. will this actually block the UI thread?  If we're getting calling on the UI thread, should we not actually kick this over to the BG thread and do the work there?

Reply:
We actually just clarify our expectations from the contract with the Editor. They asked explicitly this method to be called from the UI thread only. So, we should be fast. They can make different calls at the same time with our.

---
In reply to: [232138780](https://github.com/dotnet/roslyn/pull/29016#discussion_r232138780) [](ancestors = 232138780)

Reply:
We can't be fast though.  This is building descriptions.  It may involves lots of compiler binding and whatnot.  For a language like TS, this can involve calling out to them, which can then call out of process to their language server to respond with the info.

We very *explicitly* made sure this was async to convey this could be costly :)

We really shouldn't have anything here on the UI thread. Note: editor is free to call in on hte UI, and they can ContinueAwait(true) for their code.  But our own impl should try to relinquish the UI so that we're not blocking other stuff in VS.

Reply:
So, you mean the best way would be not assume anything here: if we are called in UI it is OK, if not in UI - not a problem. Correct?

---
In reply to: [232401945](https://github.com/dotnet/roslyn/pull/29016#discussion_r232401945) [](ancestors = 232401945)

Reply:
i'd say that if there a strong indication they'll be calling into us on the UI thread, then we shuold likely explicitly relinquish control of that thread ourselves :)

Reply:
OK. Then, let us remove the check at all. Thanks!

---
In reply to: [232404829](https://github.com/dotnet/roslyn/pull/29016#discussion_r232404829) [](ancestors = 232404829)

Reply:
This is now fixed, but as a quick clarification, asynchronous methods are not allowed to assert thread affinity. If they need to run code on a particular thread, they switch to that thread and run the code.

Reply:
Yup.  Sam summed it up much better than i did.  @sharwell One thing that would be good would be a easy doc somewhere that explains the rules.  I think i've wrapped my head around it all (including jtf), but it would be good to have so it can easily be disseminated and adopted.  Thanks!

Reply:
Actually, i take it back.  My inability to think through if AsyncLazy would work shows that i don't actually understand things totally.  So having the doc would be super helpful!

Reply:
Thank you, @sharwell! Removed asserts in async calls.

---
In reply to: [232491561](https://github.com/dotnet/roslyn/pull/29016#discussion_r232491561) [](ancestors = 232491561)

Reply:
Thank you! @sharwell, @cyrusnajmabadi, do you think it is reasonable to add this doc to ForegroundThreadAffinitizedObject?

---
In reply to: [232502417](https://github.com/dotnet/roslyn/pull/29016#discussion_r232502417) [](ancestors = 232502417)